File: recover-binary-search-tree.js
// Time Complexity : O(N)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var recoverTree = function(root) {
    if (!root) return;
    
    const stack = [];
    let first = null,
        second = null,
        prev = null;
    
    while (stack.length > 0 || root != null) {
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
        root = stack.pop();
        if (prev != null && root.val < prev.val) {
            second = root;
            if (first == null) first = prev;
            else break;
        }
        prev = root;
        root = root.right;
    }

    [first.val, second.val] = [second.val, first.val];
};

File: Recover binary search tree.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach


 // We will traverse through the tree in an inorder manner. Whenever we come across a breach , where the left node is greater than the root, 
 // or the right node is smaller than the root, then we store the value of that node in a variable. The we search for the second breach and 
 // when we find it we swap the two nodes.
 class Solution {
    TreeNode first, second, prev;

    public void recoverTree(TreeNode root) 
    {
        inorder(root);
        int temp=first.val;
        first.val=second.val;
        second.val=temp;
    }

    private void inorder(TreeNode root)
    {
        //base
        if(root==null) return;

        //logic
        inorder(root.left);
        //breach
        if(prev!=null && prev.val>root.val)
        {
            //if the first breach is not found yet
            if(first==null)
            {
                first=prev; //set first to the prev value which is greater than the root
                second=root;    //set second to root, which is the smaller one
            }
            //if first breach is found,then we look for the second one
            else
            {
                second=root;
            }
        }
        prev=root;
        inorder(root.right);
    }

}


//Iterative method
class Solution {
    TreeNode first, second, prev;

    public void recoverTree(TreeNode root) 
    {
        inorder(root);
        int temp=first.val;
        first.val=second.val;
        second.val=temp;
    }

    private void inorder(TreeNode root)
    {
        //logic
         Stack<TreeNode> st= new Stack<>();
        while(!st.isEmpty() || root!=null)
        {
            while(root!=null)
            {
                st.push(root);
                root=root.left;
            }
            root=st.pop();
        //breach
        if(prev!=null && prev.val>root.val)
        {
            //if the first breach is not found yet
            if(first==null)
            {
                first=prev; //set first to the prev value which is greater than the root
                second=root;    //set second to root, which is the smaller one
            }
            //if first breach is found,then we look for the second one
            else
            {
                second=root;
            }
        }
    
        prev=root;
        root=root.right;
        
        }
    }
}